assets/queries/terraform/aws/lambda_iam_invokefunction_misconfigured/query.rego

```diff
@@ -0,0 +1,65 @@
+package Cx
+
+import data.generic.common as common_lib
+import data.generic.terraform as tf_lib
+
+#CxPolicy for ressource iam policy
+CxPolicy[result] {
+	resource := input.document[i].resource.aws_iam_policy[name]
+
+	policy := common_lib.json_unmarshal(resource.policy)
+
+	st := common_lib.get_statement(policy)
+	statement := st[_]
+
+	check_lambda_invoke(statement) == false
+
+	result := {
+		"documentId": input.document[i].id,
+		"searchKey": sprintf("aws_iam_policy[{{%s}}]", [name]),
+		"issueType": "IncorrectValue",
+		"keyExpectedValue": sprintf("aws_iam_policy[%s].statement.ressource is misconfigured", [name]),
+		"keyActualValue": sprintf("aws_iam_policy[%s].statement.ressource allow access to function (unqualified ARN) and its sub-resources, add another statement with ":*" to function name", [name])
+	}
+}
+
+#CxPolicy for data iam policy document
+CxPolicy[result] {
+	ressource := input.document[i].data.aws_iam_policy_document[name]
+
+	check_lambda_invoke(ressource.statement[_]) == false
+
+	result := {
+		"documentId": input.document[i].id,
+		"searchKey": sprintf("aws_iam_policy[{{%s}}]", [name]),
+		"issueType": "IncorrectValue",
+		"keyExpectedValue": sprintf("aws_iam_policy[%s].statement.ressource is misconfigured", [name]),
+		"keyActualValue": sprintf("aws_iam_policy[%s].statement.ressource allow access to function (unqualified ARN) and its sub-resources, add another statement with ":*" to function name", [name]),
+	}
+}
+
+check_lambda_invoke(statement) {
+	check_iam_ressource(statement)
+	check_iam_action(statement)
+} else = false {
+	true
+}
+
+
+check_iam_ressource(statement) {
+	is_string(statement.Resource)
+	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+:[*]$)", statement.Resource)
+	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+$)", statement.Resource)
+} else {
+	is_array(statement.Resource)
+	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+:[*]$)", statement.Resource[_])
+	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+$)", statement.Resource[_])
+}
+
+check_iam_action(statement) {
+	is_string(statement.Action)
+	regex.match("(^lambda:InvokeFunction$)", statement.Action)
+} else {
+	is_array(statement.Action)
+	regex.match("(^lambda:InvokeFunction$)", statement.Action[_])
+}
```

Comment:
I think it will be also relevant to consider `aws_iam_role_policy`, `aws_iam_user_policy`, and `aws_iam_group_policy`. Furthermore, I think it will be also relevant to consider "*" in actions. What do you think?

Please, note that:

- The query is failing (not running) because of the `keyExpectedValue`. If you want to use quotes, you need to put "\" before.

- `resourceType` and `resourceName` are required in the results of policies of TF queries.

- The policy for `aws_iam_policy_document` is unnecessary since KICS scan parse data `aws_iam_policy_document` resources.

- The postive4.tf (should be renamed to positive4.tf) presents an invalid `aws_iam_policy_document`. Please, read the documentation: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document#actions

- The lines of the results you pointed in `positive_expected_result.json` are incorrect. Please run your query against your query to verify them. For example: "kics scan -p ./assets/queries/terraform/aws/lambda_iam_invokefunction_misconfigured -q 
./assets/queries/terraform/aws/lambda_iam_invokefunction_misconfigured/query.rego"

- Run `go test ./test` to verify if your query passes all the KICS queries tests

I suggest:
```suggestion
package Cx

import data.generic.common as common_lib
import data.generic.terraform as tf_lib

#CxPolicy for ressource iam policy
CxPolicy[result] {
	resourceType := {"aws_iam_role_policy", "aws_iam_user_policy", "aws_iam_group_policy", "aws_iam_policy"}

	resource := input.document[i].resource[resourceType[idx]][name]

	policy := common_lib.json_unmarshal(resource.policy)

	st := common_lib.get_statement(policy)
	statement := st[_]

	not check_lambda_invoke(statement)

	result := {
		"documentId": input.document[i].id,
		"resourceType": resourceType[idx],
		"resourceName": tf_lib.get_resource_name(resource, name),
		"searchKey": sprintf("%s[%s].policy", [resourceType[idx], name]),
		"issueType": "IncorrectValue",
		"keyExpectedValue": sprintf("%s[%s].policy is misconfigured", [name]),
		"keyActualValue": sprintf("%s[%s].policy allows access to function (unqualified ARN) and its sub-resources, add another statement with \":*\" to function name", [name])
	}

}

check_lambda_invoke(statement) {
	check_iam_ressource(statement)
	check_iam_action(statement)
}


check_iam_ressource(statement) {
	is_string(statement.Resource)
	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+:[*]$)", statement.Resource)
	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+$)", statement.Resource)
} else {
	is_array(statement.Resource)
	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+:[*]$)", statement.Resource[_])
	regex.match("(^arn:aws:lambda:.*:.*:function:[a-zA-Z0-9_-]+$)", statement.Resource[_])
}



check_iam_action(statement) {
	any([regex.match("(^lambda:InvokeFunction$)", statement.actions[_]), statement.actions[_] == "*"])
} else {
	any([regex.match("(^lambda:InvokeFunction$)", statement.Actions[_]), statement.Actions[_] == "*"])
} else {
	is_array(statement.Action)
	any([regex.match("(^lambda:InvokeFunction$)", statement.Action[_]), statement.Action[_] == "*"])
} else {
	is_string(statement.Action)
	any([regex.match("(^lambda:InvokeFunction$)", statement.Action), statement.Action == "*"])
}

```

Reply:
i agree with this change, i simplify the code. i push with a fix a see a bug when action is not lambda:invoke or "*" ressources matche :/  I add a  3 news tests cases.

Reply:
Thank you so much, @jplanckeel!